### PR TITLE
Api: Enforce bucket policy vulnerability fix

### LINKED ISSUE #1467: Api: Bucket policy vulnerability: Allowing anonymous access to protected files.
Here is how to reproduce the issue. 
- Upload file named output.txt to the bucket. 
  
  $`mc cp output.txt myminio/test1/`
- Set the bucket policy to `none`.
  
  $`mc access  none  myminio/test1/`
- Try to access `output.txt`.
  
  $`curl localhost:9000/test1/output.txt`

It fails since you don't have access to it.
- Set the bucket policy to as showed below. 
  
  $`mc access  readonly myminio/test1/oo`
- Now try 
  
  $`curl localhost:9000/test1/output.txt` 
  
  You get access to the file. 

The above policy setting shouldn't allow anonymous access to `myminio/test1/output.txt` .

This is because of the regex match [here](https://github.com/minio/minio/blob/master/bucket-policy-handlers.go#L81) .

The policy resource for object access is stored as `arn:aws:s3:::test1/oo*` and this matches `arn:aws:s3:::test1/output.txt` . 

Also it matches any object that begins with `arn:aws:s3:::test1/o` since `*` represents one or more occurrences of the character before `*` . 

Also AWS docs http://docs.aws.amazon.com/AmazonS3/latest/dev/s3-arn-format.html, specify `*` to be used as wildcard, not as regexp. 

Fix with tests in progress.

### PR DESCRIPTION
**Pull request by hackintoshrao**
Fixes #1467 .

### CODE REVIEW

**Comment by hackintoshrao** [comment]
https://github.com/minio/minio/pull/1476/files#diff-e4af08fd3fcfe7dd296515f062400768R54 is the test case validating for #1467 .

File: bucket-policy-handlers.go [hunk 1]
```diff
@@ -78,14 +78,15 @@ func bucketPolicyActionMatch(action string, statement policyStatement) bool {
...
-		matched, err := regexp.MatchString(presource, strings.TrimPrefix(resource, "/"))
-		fatalIf(err, "Invalid pattern, please verify the pattern string.", nil)
-		// For any path matches, we return quickly and the let the caller continue.
-		if matched {
-			return true
+		// the reource rule for object could contain "*" wild card.
+		// TrimSuffix removes the wildcard if it exists and check whether
+		// the requested object can be given access based on the already set bucket policy.
+		// More info: http://docs.aws.amazon.com/AmazonS3/latest/dev/s3-arn-format.html
+		if matched := strings.HasPrefix(resource, strings.TrimSuffix(presource, "*")); !matched {
```
Comment (harshavardhana) [inline bucket-policy-handlers.go]: you should use.. https://golang.org/pkg/path/#Match

Comment (hackintoshrao) [inline bucket-policy-handlers.go]: But is any other wild card thant `*` is supported by S3 protocols ? If not then there is no need to use Match .

Comment (hackintoshrao) [inline bucket-policy-handlers.go]: The problem with Match is it interprets only till `/` . We need the code snippet in the following link to be matched right ? https://play.golang.org/p/QRKlyKhBY3

File: bucket-policy-handlers.go [hunk 2]
```diff
@@ -78,14 +78,15 @@ func bucketPolicyActionMatch(action string, statement policyStatement) bool {
 // Verify if given resource matches with policy statement.
 func bucketPolicyResourceMatch(resource string, statement policyStatement) bool {
 	for _, presource := range statement.Resources {
-		matched, err := regexp.MatchString(presource, strings.TrimPrefix(resource, "/"))
-		fatalIf(err, "Invalid pattern, please verify the pattern string.", nil)
-		// For any path matches, we return quickly and the let the caller continue.
-		if matched {
-			return true
+		// the reource rule for object could contain "*" wild card.
```
Comment (harshavardhana) [inline bucket-policy-handlers.go]: 's/reource/resource'

**Pull request merged**